pkg/apic/provisioning/credentials.go

```diff
@@ -0,0 +1,98 @@
+package provisioning
+
+import "fmt"
+
+const credentialTypeSetError = "can not set credential as %s as its already set as another %s"
+
+// Credential - holds the details about the credential to send to encrypt and send to platform
+type Credential struct {
+	credentialType credentialType
+	data           interface{}
+	reference      string
+}
+
+type oAuthCredential struct {
+	id     string
+	secret string
+}
+
+type apiKeyCredential struct {
+	key string
+}
+
+// CredentialBuilder - builder to create new credentials to send to Central
+type CredentialBuilder interface {
```

Comment:
Can a credential have both oauth and api key defined?

Reply:
I'm hesitant to have a single builder for all credentials. If more credentials are added then we have to keep growing this interface and the types that implement it. Can we some kind of `Credential` interface for other hypothetical structs like `OAuthCredential` `APIKeyCredential` to implement?

Reply:
There would be separate requests for each type, but this builder will support all credential types. If they try to set 2 different types an error is returned

Reply:
Based on the tests it doesn't look like we will be setting both types of credentials at once, so I don't think a builder is the right approach here.

Reply:
Something like this to break them up and keep each credential independent of the other

```
type credentialType int

const (
	// credentialTypeOAuth - OAuth credentials
	credentialTypeOAuth credentialType = iota + 1
	// credentialTypeAPIKey - APIKey credentials
	credentialTypeAPIKey
)

type Credential struct {
	credentialType credentialType
	data           interface{}
	reference      string
}

type oAuthCredential struct {
	id     string
	secret string
}

type apiKeyCredential struct {
	key string
}

func NewOAuthCredential(id, secret, ref string) Credential {
	c := Credential{
		credentialType: credentialTypeOAuth,
		reference:      ref,
	}
	c.data = &oAuthCredential{
		id:     id,
		secret: secret,
	}

	return c
}

func NewAPIKeyCredential(key, ref string) Credential {
	c := Credential{
		credentialType: credentialTypeAPIKey,
		reference:      ref,
	}
	c.data = &apiKeyCredential{
		key: key,
	}

	return c
}
```

Reply:
How will the agents be creating a credential? I see it will return an error, but I think the builder adds more complexity than having two separate types.

Reply:
The builders allow us greater flexibility for changes in the SDK in the future. Thinking for this we may need to add additional setter methods for the API Key to say if it's in the headers or query params. With the builder we can add that additional method without causing issues on the agent implementation.